demos/src/transport_plaintext.c

```diff
@@ -0,0 +1,158 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "transport_plaintext.h"
+
+#include <netdb.h>
+#include <unistd.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+int Transport_ConnectToServer( const char * pServer, uint16_t port )
+{
+    int status, tcpSocket = -1;
+    struct addrinfo * pListHead = NULL, * pIndex;
+    struct sockaddr * pServerInfo;
+    uint16_t netPort = htons( port );
+    socklen_t serverInfoLength;
+
+    status = getaddrinfo( pServer, NULL, NULL, &pListHead );
+
+    if( status != -1 )
+    {
+        for( pIndex = pListHead; pIndex != NULL; pIndex = pIndex->ai_next )
+        {
+            tcpSocket = socket( pIndex->ai_family, pIndex->ai_socktype, pIndex->ai_protocol );
+
+            if( tcpSocket == -1 )
+            {
+                continue;
+            }
+
+            pServerInfo = pIndex->ai_addr;
+
+            if( pServerInfo->sa_family == AF_INET )
+            {
+                /* IPv4 */
+                ( ( struct sockaddr_in * ) pServerInfo )->sin_port = netPort;
+                serverInfoLength = sizeof( struct sockaddr_in );
+            }
+            else
+            {
+                /* IPv6 */
+                ( ( struct sockaddr_in6 * ) pServerInfo )->sin6_port = netPort;
+                serverInfoLength = sizeof( struct sockaddr_in6 );
+            }
+
+            status = connect( tcpSocket, pServerInfo, serverInfoLength );
+
+            if( status == -1 )
+            {
+                close( tcpSocket );
+            }
+            else
+            {
+                break;
+            }
+        }
+
+        if( pIndex == NULL )
+        {
+            status = -1;
+        }
+        else
+        {
+            status = tcpSocket;
+        }
+    }
+
+    if( pListHead != NULL )
+    {
+        freeaddrinfo( pListHead );
+    }
+
+    return status;
+}
+
+int32_t Transport_SendExact( int tcpSocket, const void * pMessage, size_t bytesToSend )
+{
+    const uint8_t * pIndex = pMessage;
+    size_t bytesRemaining = bytesToSend;
+    int32_t totalBytesSent = 0;
+    ssize_t bytesSent;
+
+    while( bytesRemaining > 0 )
+    {
+        bytesSent = send( tcpSocket, pIndex, bytesRemaining, 0 );
+
+        if( bytesSent > 0 )
+        {
+            bytesRemaining -= ( size_t ) bytesSent;
+            totalBytesSent += ( int32_t ) bytesSent;
+            pIndex += bytesSent;
+        }
+        else
+        {
+            totalBytesSent = -1;
+            break;
+        }
+    }
+
+    return totalBytesSent;
+}
+
+int32_t Transport_RecvExact( int tcpSocket, void * pBuffer, size_t bytesToRecv )
```

Comment:
Are you expecting the MQTT demo to use this and the HTTP to use Transport_RecvUpTo()?
If so, then that would mean that MQTT and HTTP have different expectations from their *recv* interface. I thought we did not want this?

Reply:
Yes that's the idea.

Reply:
Received news that how to handle partial reads in MQTT is still in discussion.

Reply:
Yes, that's why this is not handling it and requires the full length to be received.

Reply:
The design calls for transport_read and transport_write. So MQTT library must only use one Transport_Recv() and not both read functions.  If MQTT decides to support partial reads, then MQTT library will have to use 'upto' function.

Reply:
The MQTT library right now is currently only using SendExact and RecvExact.

Reply:
After having transport layer discussions, it was decided that the library receive loop should handle partial reads. So RecvExact  is not needed. In most of the cases, receive will behave like RecvExact, but the loop will need to handle the case when not all bytes are received.